core/templates/dev/head/pages/exploration_editor/EditorServices.js

```diff
@@ -50,12 +50,16 @@ oppia.factory('explorationData', [
     var explorationData = {
       explorationId: explorationId,
       autosaveChangeList: function(changeList, successCallback, errorCallback) {
+        // First save locally to be retrieved later if save is unsuccessful.
+        localSaveService.saveExplorationDraft(explorationId, changeList,
+          draftChangeListId);
         $http.put(explorationDraftAutosaveUrl, {
           change_list: changeList,
           version: explorationData.data.version
         }).then(function(response) {
+          draftChangeListId = response.data.draft_change_list_id;
+          localSaveService.removeExplorationDraft(explorationId);
           if (successCallback) {
-            draftChangeListId = response.data.draft_change_list_id;
             successCallback(response);
           }
         }, function() {
```

Comment:
Oh yes, I suggested implementing milestone 4 in this PR. I think it makes sense to do it here, yeah. Then you can test the end-to-end flow before ever saving anything in production.

Thanks!